File: Java/Graph/Clone_Graph.java
// 133. Clone Graph
// Leetcode Medium
// Date: 8 April 2023
// TC: O(N)
// SC: O(N)


// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}


class Solution {
    public Node clone(Node node, HashMap<Node, Node> vis )
    {
        Node newNode = new Node(node.val);
        vis.put(node, newNode);

        for(Node nod : node.neighbors)
        {
            if(vis.containsKey(nod))
                newNode.neighbors.add(vis.get(nod));
            else
                newNode.neighbors.add(clone(nod, vis));
        }
        return newNode;
    }
    public Node cloneGraph(Node node) {
        if (node == null) return null;
        HashMap<Node, Node> vis = new HashMap<Node, Node>();
        return clone(node, vis);

    }
}

File: LeetCode/DailyCodingChallenge/April 2023/Clone_Graph.py
# 133. Clone Graph
# Leetcode Medium
# Date: 8 April 2023
# TC: O(N)
# SC: O(N)


# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def dfs(self, node, vis):
        new_node = Node(node.val)
        vis[node.val] = new_node
        for nod in node.neighbors:
            if nod.val in vis:
                new_node.neighbors.append(vis[nod.val])
            else:
                new_node.neighbors.append(self.dfs(nod, vis))
        return new_node

    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node: return
        vis = {}

        return self.dfs(node, vis)
